File: typescript/packages/bench/src/support/abstractMetric.ts
/** Base shared by all 16 metrics. */

import { MetricUnit, type MetricContract } from "../contracts/metricContract";
import type { Dataset } from "../dataset";
import type { HasLogging } from "../concerns/hasLogging";

export abstract class AbstractMetric implements MetricContract, HasLogging {
    protected constructor(
        readonly key: string,
        readonly unit: MetricUnit,
    ) {}

    abstract compute(data: Dataset): number;

    formatted(data: Dataset): string {
        return `${this.compute(data).toFixed(2)}${MetricUnit.suffix(this.unit)}`;
    }

    logTarget(): string {
        return this.key;
    }

    logLine(message: string): string {
        return `[${this.logTarget()}] ${message}`;
    }
}

File: rust/crates/atelier-bench/src/support/abstract_metric.rs
//! Base shared by all 16 metrics.

use crate::concerns::has_logging::HasLogging;
use crate::contracts::metric_contract::MetricUnit;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct MetricBase {
    pub key: &'static str,
    pub unit: MetricUnit,
}

impl MetricBase {
    pub const fn new(key: &'static str, unit: MetricUnit) -> Self {
        Self { key, unit }
    }
}

impl HasLogging for MetricBase {
    fn log_target(&self) -> &'static str {
        self.key
    }
}
